lock/cond.go

```diff
@@ -0,0 +1,162 @@
+package lock
+
+import (
+	"container/list"
+	"context"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+type Ticket interface {
+	// Release releases the ticket. The ticket must not be released multiple times.
+	Release()
+	// Wait waits for someone to wakeup the context. It is not concurrency safe [in that multiple goroutines cannot
+	// wait on the same ticket]. If the condition is woken up, no error is returned, the ticket remains valid.
+	// The ticket now holds an exclusive lock on the condition, and it must be released for others to
+	// make progress.
+	// Otherwise, ctx.Err() is returned by Wait, and the ticket is rendered invalid.
+	Wait(context.Context) error
```

Comment:
It seems that `Wait` and `Waiter` are exclusive in the implementation.
I'm not sure in what case another package would want to call `Waiter()` and seems like `Waiter` does not need to be exposed as an interface?

Reply:
Waiter can be used for 1:N style broadcast with channels. This isn't used right now, but it's used in a future controller (When we start doing multiple controllers listening to a change.

Reply:
I removed the interface.